### PR TITLE
[Controlador Usuario] - Create user controller

### PR DESCRIPTION
**Pull request by AleFalcon**
## Summary
Create user controller with views in html5

## Trello Card
https://trello.com/c/Z5wPj4WA/21-crear-user-controller-con-las-operaciones-crud-de-user

### CODE REVIEW

File: src/main/java/wolox/training/models/User.java [hunk 1]
```diff
@@ -41,38 +43,40 @@
     )
     private List<Book> books;
 
+    @Autowired
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java]: This is not necessary

File: src/main/java/wolox/training/models/User.java [hunk 2]
```diff
@@ -41,38 +43,40 @@
...
     public User() {
         this.setBooks(null);
     }
 
-    public User(String username, String name, LocalDate birthday) {
+    @Autowired
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/User.java]: This is not necessary

File: src/main/java/wolox/training/models/Book.java [hunk 3]
```diff
@@ -158,4 +158,8 @@ public void update (String genre, String author, String image, String title, Str
         this.setYear(year);
         this.setPage(page);
     }
+
+    public void add_user(User user){
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/models/Book.java]: I don't think this method will be used. Once a book is automatically acquired by a user with the declared relationship, the user will be added to the book list.

File: src/main/java/wolox/training/errors/user/UserNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package wolox.training.errors.user;
+
+public class UserNotFoundException extends RuntimeException {
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/errors/user/UserNotFoundException.java]: I could add this exception to the existing error handler.

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    public User update(@RequestBody User user) {
+        User found_user = found_user(user.getUsername(), userRepository);
+        found_user.update(user.getName(), user.getBirthday());
+        userRepository.save(found_user);
+
+        return found_user;
+    }
+
+    @DeleteMapping("/delete")
+    public void delete(@RequestBody User user) {
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java]: If your search method uses username only, I don't think sending the entire book is necessary. You can send the username as @PathVariable or @RequestParam.

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        User found_user = found_user(user.getUsername(), userRepository);
+        userRepository.delete(found_user);
+    }
+
+    @GetMapping("/details")
+    public User read(@RequestBody User user) {
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java]: The same as in the delete method.

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        return found_user(user.getUsername(), userRepository);
+    }
+
+    @GetMapping("/list_books")
+    public List<Book> list_book(@RequestBody User user) {
+        return found_user(user.getUsername(), userRepository).getListBooks();
+    }
+
+    @PutMapping("/add_book")
+    public User add_book(@RequestBody User user, @RequestBody Book book) {
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java]: The same as in the delete method.

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+public class UserApiController extends ApiController {
+
+    @Autowired
+    private UserRepository userRepository;
+    @Autowired
+    private BookRepository bookRepository;
+
+    @PostMapping("/create")
+    @ResponseStatus(HttpStatus.CREATED)
+    public User create(@RequestBody User user) {
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java:22]: t is considered a good practice in Api Rest not to create specific routes for each method, that is, not to have a path "api/users/create", "api/users/delete", etc.

It is enough to have the different http method to differentiate one url from another. With the same http method you are saying if it is a cration, a search, an update or a deletion, it is not necessary to write it in the url.

I recommend using the following URLs:

POST api/users

DELETE api/users

GET api/ users/{username}

PUT api/books/{username}

**Review by Gianni-Nicolas — CHANGES_REQUESTED** [review]
Great job

File: src/main/java/wolox/training/controllers/UserApiController.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+        return found_user;
+    }
+
+    @DeleteMapping
+    public void delete(@RequestParam(name="username", required=true) String username) {
+        userRepository.delete(found_user(username, userRepository));
+    }
+
+    @GetMapping("{username}")
```
Comment (Gianni-Nicolas) [inline src/main/java/wolox/training/controllers/UserApiController.java]:
```suggestion
    @GetMapping("/{username}")
```

**Review by Gianni-Nicolas — CHANGES_REQUESTED** [review]

**Review by Gianni-Nicolas — DISMISSED** [review]

**Review by facucompe — DISMISSED** [review]
Great work

**Review by facucompe — APPROVED** [review]

**Review by Gianni-Nicolas — APPROVED** [review]

**Pull request merged**